server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -260,13 +268,19 @@ public DesiredBalance compute(
                 break;
             }
             if (i % 100 == 0) {
-                // TODO this warning should be time based, iteration count should be proportional to the number of shards
+                final int iterations = i;
+                final long currentTime = System.nanoTime();
```

Comment:
I'd prefer to use `ThreadPool#relativeTimeInMillis` here. I'm sure there's a threadpool lying around somewhere nearby...